### PR TITLE
[#12048] Migrate Get Student and Students Action

### PR DESCRIPTION
**Pull request by domlimm**
Part of #12048.

**Things to highlight**

Not sure if it is the best way to do this (similar to PR #12210):

```java
EntitysData data = new EntitysData();
List<EntityData> entityDataList = entitysForCourse
        .stream()
        .map(EntityData::new)
        .collect(Collectors.toList());

data.setEntitys(entityDataList);
```

As EntitysData class current constructor with EntityAttribute will cause a type erasure problem when adding another constructor taking in a List of different data type.

### CODE REVIEW

File: src/main/java/teammates/storage/sqlapi/UsersDb.java [hunk 1]
```diff
@@ -140,6 +141,25 @@ public Student getStudentByGoogleId(String courseId, String googleId) {
...
 
+    /**
+     * Gets a list of students by {@code teamName} and {@code courseId}.
+     */
+    public List<Student> getStudentsByTeamName(String teamName, String courseId) {
+        CriteriaBuilder cb = HibernateUtil.getCriteriaBuilder();
+        CriteriaQuery<Student> cr = cb.createQuery(Student.class);
+        Root<Student> studentRoot = cr.from(Student.class);
+
+        studentRoot.alias("student");
```
Comment (samuelfangjw) [inline src/main/java/teammates/storage/sqlapi/UsersDb.java:152]: Why is this needed?

File: src/main/java/teammates/ui/webapi/GetStudentAction.java [hunk 2]
```diff
@@ -25,64 +28,129 @@ AuthType getMinAuthLevel() {
...
+                studentData.setGoogleId(student.getAccount().getGoogleId());
+            }
 
-        if (studentEmail == null) {
-            student = getPossiblyUnregisteredStudent(courseId);
+            if (studentEmail == null) {
+                // hide information if not an instructor
+                studentData.hideInformationForStudent();
+                // add student institute
+                studentData.setInstitute(sqlLogic.getCourse(courseId).getInstitute());
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/GetStudentAction.java:121]: student.getCourse().getInstitute()

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/storage/sqlapi/UsersDb.java (hunk 1 above)
Comment (domlimm) [reply src/main/java/teammates/storage/sqlapi/UsersDb.java:152]: Team and User have duplicated attribute `name`. Curious from the previous comment on commenting, should I have commented for this?

File: src/it/java/teammates/it/ui/webapi/GetStudentActionIT.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import teammates.ui.webapi.JsonResult;
+
+/**
+ * SUT: {@link GetStudentAction}.
+ */
+public class GetStudentActionIT extends BaseActionIT<GetStudentAction> {
+
+    @Override
+    @BeforeMethod
+    protected void setUp() throws Exception {
```
Comment (cbell504) [inline src/it/java/teammates/it/ui/webapi/GetStudentActionIT.java:23]: What would be the purpose to throw a generic exception here and a few other places instead of something more specific?

File: src/main/java/teammates/ui/output/StudentData.java [hunk 4]
```diff
@@ -37,6 +38,16 @@ public StudentData(StudentAttributes studentAttributes) {
...
     }
 
+    public StudentData(Student student) {
+        this.email = student.getEmail();
+        this.courseId = student.getCourseId();
+        this.name = student.getName();
+        this.joinState = student.isRegistered() ? JoinState.JOINED : JoinState.NOT_JOINED;
+        this.comments = student.getComments();
+        this.teamName = student.getTeam().getName();
+        this.sectionName = student.getTeam().getSection().getName();
```
Comment (cbell504) [inline src/main/java/teammates/ui/output/StudentData.java:48]: Is there any chance that getTeam() or getSection() could return null?

**Comment by samuelfangjw** [comment]
@cbell504, thanks for the review, but would appreciate if you left reviews (particularly for the v9 project) to the project maintainers.

To answer your two queries, generic exception is thrown because this is a test method. The only desired behaviour is that the test runner catches the exception and fails the test  when an unexpected exception is thrown. None of the usual arguments for throwing specific exceptions in production code apply here. In fact, throwing multiple specific exceptions here only adds noise and decreases readability.

For the second issue, other than the Law of Demeter violation, no major issues with getTeam and getSection here. These should never return null because it's a business requirement that these are never null for students. Any null values is indicative of a programmer error.

File: src/main/java/teammates/ui/webapi/GetStudentsAction.java [hunk 5]
```diff
@@ -42,34 +61,94 @@ void checkSpecificAccessControl() throws UnauthorizedAccessException {
...
+                    }
+                });
+
+                StudentsData data = new StudentsData();
+                List<StudentData> studentDataList = studentsToReturn
+                        .stream()
+                        .map(StudentData::new)
+                        .collect(Collectors.toList());
+
+                data.setStudents(studentDataList);
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/GetStudentsAction.java:104]: Same here, the migrated entities should use the constructor. The datastore entities can use the workaround method.

Comment (domlimm) [reply src/main/java/teammates/ui/webapi/GetStudentsAction.java]: Sorry I miss that out..!

**Review by samuelfangjw — APPROVED** [review]

**Comment by samuelfangjw** [comment]
@domlimm Kindly help to resolve conflicts, thank you!
